### PR TITLE
Terragrunt performance improvements

### PR DESCRIPTION
**Pull request by denis256**
## Description

A couple of performance improvements by using in-memory cache of git execution and parsing of dependencies

Example runs on my test repos:
output

Included changes:
* added caching of HCL parsing struct
* added caching of `git --show-toplevel`
* updated fetching of TerragruntConfig to do full clone of struct
* replaced global variables cache with data from context
* added metrics for cache miss/hits

Close: #2344

References: #2980

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

### CODE REVIEW

**Review by yhakbar — APPROVED** [review]
LGTM. Great Job!

File: config/cache_test.go [hunk 1]
```diff
@@ -23,9 +24,10 @@ func TestTerragruntConfigCacheOperation(t *testing.T) {
 
 	testCacheKey := "super-safe-cache-key"
 
-	cache := cache.NewCache[TerragruntConfig]()
+	ctx := context.Background()
+	cache := cache.NewCache[TerragruntConfig]("TerragruntConfig")
```
Comment (levkohimins) [inline config/cache_test.go:28]: I would use constants everywhere, for example we could reuse them in unit tests.
```suggestion
	cache := cache.NewCache[TerragruntConfig](TerragruntConfigCacheName)
```

File: config/context.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	"github.com/gruntwork-io/terragrunt/config/hclparse"
+	"github.com/gruntwork-io/terragrunt/internal/cache"
+)
+
+const (
+	HclCacheContextKey              configKey = iota
+	TerragruntConfigCacheContextKey configKey = iota
+	RunCmdCacheContextKey           configKey = iota
+	DependencyOutputCacheContextKey configKey = iota
+)
```
Comment (levkohimins) [inline config/context.go:15]:
```suggestion
type contextKey byte

const (
	HclCacheContextKey contextKey = iota
	TerragruntConfigCacheContextKey
	RunCmdCacheContextKey
	DependencyOutputCacheContextKey
)
```

File: internal/cache/cache.go [hunk 3]
```diff
@@ -1,40 +1,52 @@
...
 	"time"
+
+	"github.com/gruntwork-io/terragrunt/telemetry"
 )
 
 // Cache - generic cache implementation
 type Cache[V any] struct {
+	Name  string
 	Cache map[string]V
 	Mutex *sync.Mutex
```
Comment (levkohimins) [inline internal/cache/cache.go:17]:
```suggestion
	Mutex *sync.RWMutex
```

and then:
```go
func (c *Cache[V]) Get(ctx context.Context, key string) (V, bool) {
	c.Mutex.RLock()
	defer c.Mutex.RUnlock()
```

```go
func (c *Cache[V]) Put(ctx context.Context, key string, value V) {
	c.Mutex.Lock()
	defer c.Mutex.Unlock()
```

File: util/reflect.go [hunk 4]
```diff
@@ -62,3 +62,84 @@ func MustWalkTerraformOutput(value interface{}, path ...string) interface{} {
 	}
 	return found
 }
+
```
Comment (levkohimins) [inline util/reflect.go:65]: Would you consider using third party packages that cover more types? This way we wouldn't have problems when adding new types to `TerragruntConfig` in the future.

https://github.com/barkimedes/go-deepcopy
https://github.com/huandu/go-clone

**Review by levkohimins — APPROVED** [review]
Good job! I left some comments, what to do with them is up to you.

File: config/cache_test.go (hunk 1 above)
Comment (levkohimins) [reply config/cache_test.go]: I meant to use the constant 

```go
const  TerragruntConfigCacheName = "TerragruntConfig"
```

**Review by yhakbar — DISMISSED** [review]

**Review by levkohimins — APPROVED** [review]

**Pull request merged**